pkg/gardenlet/controller/managedseed/managedseed_actuator.go

```diff
@@ -0,0 +1,835 @@
+
+	// Remove managed seed reference from the shoot
+	if shoot.Spec.ManagedSeedName != nil {
+		managedSeedLogger.Infof("Removing managed seed reference from shoot %s", kutil.ObjectName(shoot))
+		shoot.Spec.ManagedSeedName = nil
+		if err := a.gardenClient.Client().Update(ctx, shoot); err != nil {
+			return fmt.Errorf("could not remove managed seed reference from shoot %q: %+v", kutil.ObjectName(shoot), err)
+		}
+	}
+
+	// Deregister the shoot as seed
+	managedSeedLogger.Infof("Deregistering shoot %s as seed", kutil.ObjectName(shoot))
+	if err := a.deregisterAsSeed(ctx, managedSeed, shoot, managedSeed.Spec.Seed != nil); err != nil {
+		return fmt.Errorf("could not dereigster shoot %q as seed: %+v", kutil.ObjectName(shoot), err)
+	}
+
+	if managedSeed.Spec.Gardenlet != nil {
+		// Delete gardenlet from the shoot
+		managedSeedLogger.Infof("Deleting gardenlet from shoot %s", kutil.ObjectName(shoot))
+		if err := a.deleteGardenlet(ctx, managedSeed, shoot); err != nil {
+			return fmt.Errorf("could not delete gardenlet from shoot %q: %+v", kutil.ObjectName(shoot), err)
+		}
+	}
+
+	// Delete garden namespace from the shoot
+	managedSeedLogger.Infof("Deleting garden namespace from shoot %s", kutil.ObjectName(shoot))
+	gardenNamespace := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: v1beta1constants.GardenNamespace}}
+	if err := a.shootClient.Client().Delete(ctx, gardenNamespace); client.IgnoreNotFound(err) != nil {
+		return fmt.Errorf("could not delete garden namespace from shoot %q: %+v", kutil.ObjectName(shoot), err)
+	}
+
+	return nil
+}
+
+func (a *actuator) registerAsSeed(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) error {
+	// Check if the shoot can be registered as seed
+	if shoot.Spec.DNS == nil || shoot.Spec.DNS.Domain == nil {
+		return errors.New("cannot register shoot as seed as it does not specify a domain")
+	}
+
+	// Prepare seed spec
+	seedSpec, err := a.prepareSeedSpec(ctx, managedSeed, shoot, &managedSeed.Spec.Seed.Spec)
+	if err != nil {
+		return err
+	}
+
+	// Create or update seed object
+	seed := &gardencorev1beta1.Seed{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: managedSeed.Name,
+		},
+	}
+	_, err = controllerutil.CreateOrUpdate(ctx, a.gardenClient.Client(), seed, func() error {
+		seed.OwnerReferences = []metav1.OwnerReference{
+			*metav1.NewControllerRef(managedSeed, seedmanagementv1alpha1.SchemeGroupVersion.WithKind("ManagedSeed")),
+		}
+		seed.Labels = utils.MergeStringMaps(managedSeed.Spec.Seed.Labels, map[string]string{
+			v1beta1constants.GardenRole: v1beta1constants.GardenRoleSeed,
+		})
+		seed.Annotations = managedSeed.Spec.Seed.Annotations
+		seed.Spec = *seedSpec
+		return nil
+	})
+	return err
+}
+
+func (a *actuator) deregisterAsSeed(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot, checkOwner bool) error {
```

Comment:
Unused params
```suggestion
func (a *actuator) deregisterAsSeed(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed) error {
```

Reply:
OK, leftover from an earlier version I forgot to remove, sorry